server/src/test/java/org/elasticsearch/search/profile/AbstractProfileBreakdownTests.java

```diff
@@ -0,0 +1,150 @@
+        assertThat(breakdownMap.get(TestTimingTypes.THREE.name()), equalTo(0L));
+        assertThat(breakdownMap.get(TestTimingTypes.THREE.name() + "_count"), equalTo(0L));
+
+        Timer firstTimerThree = testBreakdown.getNewTimer(TestTimingTypes.THREE);
+        int firstTimerThreeCount = randomIntBetween(10, 20);
+        runTimerNTimes(firstTimerThree, firstTimerThreeCount);
+
+        breakdownMap = testBreakdown.toBreakdownMap();
+        assertThat(breakdownMap.size(), equalTo(6));
+        assertThat(breakdownMap.get(TestTimingTypes.ONE.name()), equalTo(firstTimerOne.getApproximateTiming()));
+        assertThat(breakdownMap.get(TestTimingTypes.ONE.name() + "_count"), equalTo(firstTimerOne.getCount()));
+        assertThat(
+            breakdownMap.get(TestTimingTypes.TWO.name()),
+            equalTo(firstTimerTwo.getApproximateTiming() + secondTimerTwo.getApproximateTiming())
+        );
+        assertThat(breakdownMap.get(TestTimingTypes.TWO.name() + "_count"), equalTo(firstTimerTwo.getCount() + secondTimerTwo.getCount()));
+        assertThat(breakdownMap.get(TestTimingTypes.THREE.name()), equalTo(firstTimerThree.getApproximateTiming()));
+        assertThat(breakdownMap.get(TestTimingTypes.THREE.name() + "_count"), equalTo(firstTimerThree.getCount()));
+
+        assertThat(
+            testBreakdown.toNodeTime(),
+            equalTo(
+                firstTimerOne.getApproximateTiming() + firstTimerTwo.getApproximateTiming() + secondTimerTwo.getApproximateTiming()
+                    + firstTimerThree.getApproximateTiming()
+            )
+        );
+    }
+
+    public void testMultiThreaded() throws InterruptedException {
+        TestProfileBreakdown testBreakdown = new TestProfileBreakdown();
+        Thread[] threads = new Thread[200];
+        final CountDownLatch latch = new CountDownLatch(1);
+        int startsPerThread = between(1, 5);
+        for (int t = 0; t < threads.length; t++) {
+            final TestTimingTypes timingType = randomFrom(TestTimingTypes.values());
+            threads[t] = new Thread(() -> {
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                Timer timer = testBreakdown.getNewTimer(timingType);
+                for (int runs = 0; runs < startsPerThread; runs++) {
+                    timer.start();
+                    timer.stop();
+                }
+            });
+            threads[t].start();
+        }
+        latch.countDown();
+        for (Thread t : threads) {
+            t.join();
+        }
+        Map<String, Long> breakdownMap = testBreakdown.toBreakdownMap();
+        long totalCounter = breakdownMap.get(TestTimingTypes.ONE + "_count") + breakdownMap.get(TestTimingTypes.TWO + "_count")
+            + breakdownMap.get(TestTimingTypes.THREE + "_count");
+        assertEquals(threads.length * startsPerThread, totalCounter);
```

Comment:
should we also check that all timers have been stopped?

Reply:
I see we already call the timers "getCount" method when returning the breakdown map, this asserts that started timers also have been stopped (https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/profile/Timer.java#L63)